src/background_iterator.rs

```diff
@@ -0,0 +1,54 @@
+use std;
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread;
+
+/// Execute an iterator on a worker thread, which can work ahead a configurable number of items.
+pub struct BackgroundIterator<T> {
+    rx: Receiver<Option<T>>,
+    done: bool,
+}
+
+impl<T: Send> Iterator for BackgroundIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        if self.done {
+            return None;
+        }
+
+        match self.rx.recv() {
+            Ok(Some(v)) => Some(v),
+            Ok(None) => {
+                self.done = true;
+                None
+            }
+
+            // if the producer thread dies, the iterator
+            // silently exits, without an error.
+            Err(_) => None,
+        }
+    }
+}
+
+impl<T: 'static + Send> BackgroundIterator<T> {
+    /// Iterate through `itr` on a newly created thread, and send items back to the returned
+    /// `BackgroundIterator` for consumption on the calling thread. The worker thread will
+    /// continue to produce elements until it is `max_read_ahead` items ahead of the consumer iterator.
+    pub fn new<I: 'static + Send + Iterator<Item = T>>(
+        itr: I,
+        max_read_ahead: usize,
+    ) -> BackgroundIterator<T> {
+        let (tx, rx) = sync_channel::<Option<T>>(max_read_ahead);
+        let _ = thread::spawn(move || {
```

Comment:
Shouldn't something be joining on this thread when the `BackgroundIterator` is destroyed?  Or would that cause a deadlock if the receiver calling thread fails because it can't destroy the `BackgroundIterator` until it's joined the thread but the sender won't die until the receiver has been destroyed?  I think it's probably fine as is but it does feel a little dirty not to be keeping track of that thread somewhere.

Reply:
I think it is more correct now. We do the join if we get None, or if we get an `Err` from `recv()`. (I think the send could panic during its drop, after sending the last `None`,  so we should presumably propagate such a panic?)